working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,924 @@
+matches any object and binds it to `a`. The pattern `(a)` invokes `field0` on
+the object and binds the result of *that* to `a`.
+
+**TODO: If this is a concern, we could avoid this weird corner by making it a
+syntax error for a field matcher to only have a single positional field. Should
+we?**
+
+### Extractor patterns
+
+An **extractor pattern** is similar but has an identifier (possibly qualified)
+before the parenthesized subpattern list. This name is used to lookup a
+extractor function that is applied to the matched value. The result of that
+function is then the object whose getters are accessed.
+
+For example:
+
+```dart
+class Color {
+  int r, g, b;
+
+  static (int, int, int) hsv(Color color) {
+    var h = calculateHue(color);
+    var s = calculateSaturation(color);
+    var v = calculateValue(color);
+    return (h, s, v);
+  }
+}
+
+printHsv(Color c) {
+  var hsv(h, s, v) = c; // <--
+  print("$h $s $v");
+}
+```
+
+On the marked line, the `hsv(h, s, v)` pattern first passes the matched value to
+`Color.hsv()`. That returns a tuple, which the pattern then destructures using
+its `(h, s, v)` subpatterns. You can think of extractors as the mirror to
+functions&mdash;where a function gives users a way to introduce a new named
+expression that performs arbitrary computation, extractors introduce a new named
+pattern.
+
+If the declared return type of the extractor is non-nullable, then the extractor
+can be used in both binder and matcher patterns. If it is nullable, then
+returning null means that the passed in value fails to match the pattern. These
+extractors can only be used in matchers.
+
+**TODO: What about potentially nullable return types like `T extract<T>()`?**
+
+Refutable extractors like this can be useful for implementing sum type style
+class hierarchies. In the previous example:
+
+```dart
+printGeometry(Geometry geometry) {
+  match (geometry) {
+    case Point(var x, var y) => print("Point $y, $y");
+    case Circle(var x, var y, var radius) => print("Circle $x, $y, $radius");
+  }
+}
+```
+
+Here, `Point()` is a extractor defined something like:
+
+```dart
+class Point extends Shape {
+  static Point? extract(Shape shape) => shape is Point ? shape : null;
+}
+```
+
+**TODO: Is it necessary to define explicit extractors for downcasts? Should we
+consider allowing `is` before a extractor to do a cast check?**
+
+### Resolving extractor functions
+
+We resolve the identifier in a extractor pattern to a function like so:
+
+1.  If the name is unqualified and there is a static method defined on the
+    value's static type with that name, the extractor is that method. *This lets
+    a class define its own "named extractors" that take precedence over the
+    lexical scope as in the `hsv()` example. Making these instance or extension
+    members on the class lets them have short names without risk of top level
+    collision.*
+
+3.  Otherwise, if the name resolves to a top level function, that function is
```

Comment:
Do you include extensions here?  You mention them in various places but none of these cases admit them that I can see.